FEATURE: Asset Sources support for Asset Editor

robertlemke opened the pull request:
This change introduces the necessary backend code for enabling the Asset Editor (in the user interface's inspector panel) to use asset sources. A user can now search for assets which are stored in asset sources and import them on the fly.

Resolves #2198

robertlemke commented:
See also https://github.com/neos/neos-ui/pull/2161

robertlemke commented:
This change should be ready to merge, unless you find some issues of course. It does not harm backwards-compatibility with the EmberJS UI, but doesn't support it.

robertlemke commented:
I guess we should target this to 4.x and release it as part of our bugfix releases. @kdambekalns / @kitsunet what do you think? And 4.0 or just 4.1?

daniellienert commented on Neos.Media/Classes/Domain/Service/AssetSourceService.php:
```diff
@@ -0,0 +1,179 @@
+     * @throws AssetSourceServiceException
+     * @throws \Exception
+     */
+    public function importAsset(string $assetSourceIdentifier, string $assetIdentifier)
```
Return type and annotation should be added

robertlemke replied:
Return type, sure! But which annotation is missing?

daniellienert commented on Neos.Neos/Classes/Controller/Service/AssetProxiesController.php:
```diff
@@ -0,0 +1,161 @@
+     * @param string $assetSourceIdentifier If specified, results are only from the given asset source
+     * @return string
+     */
+    public function indexAction($searchTerm = '', $assetSourceIdentifier = '')
```
Parameter types  should be added

daniellienert commented:
As far as I understood it, this is a behavior changing feature which even could break the asset selection and should NOT go into a bugfix release.

robertlemke commented on Neos.Neos/Classes/Controller/Service/AssetProxiesController.php:
```diff
@@ -0,0 +1,161 @@
+    {
+        $assetSources = $this->assetSourceService->getAssetsSources();
+        $assetProxies = [];
+        foreach ($assetSources as $assetSource) {
```
Yes, agreed, I'm going to add that functionality.

bwaidelich replied:
Is this resolved?

robertlemke replied:
Yes, it is resolved.

robertlemke commented:
> As far as I understood it, this is a behavior changing feature which even could break the asset selection and should NOT go into a bugfix release.

This should not break any existing behavior, if it does, I need to fix that. For Neos setups which don't have asset sources, everything should work as without this patch. And Neos setups which have asset sources configured, should just see the additional support in the asset editor.

Technically, we haven't officially released the asset sources API yet, so if that needs to be changed (which I don't expect), it would be non-breaking as well.

robertlemke commented:
I added the missing type declarations and additional doc comments.

Also added a limit parameter for the asset proxies web service endpoint. The limit is per asset source and can be provided as a GET parameter by the user interface.

daniellienert commented:
@robertlemke, thanks for the changes! With that it will not harm our projects. Still its a behavior changing feature which should go into a minor release at least.

bwaidelich commented:
@daniellienert I didn't look into this yet, but from the description it should just have an effect on the asset editors. Can you elaborate which behavioral changes you are referring to?

bwaidelich commented on Neos.Neos/Classes/Controller/Service/AssetProxiesController.php:
```diff
@@ -0,0 +1,167 @@
+            }
+            try {
+                $assetProxyRepository = $assetSource->getAssetProxyRepository();
+                $assetProxyQueryResult = $assetProxyRepository->findBySearchTerm($searchTerm);
```
I guess this one needs a limit, too!

robertlemke replied:
Well, that's implemented in line 108.

bwaidelich replied:
What I meant is: We need to limit that on the asset source's side. Otherwise an asset source could download all images from Wikipedia only to be limited afterwards :)

robertlemke replied:
IMO, that is the responsibility of the asset source. The interface says that it should return a query result with Iterator support. What I do in my asset source is, that I fetch x asset proxies at a time and if you iterate further, fetch the next ones. Not all asset sources can support something like limit or offset, because they simply have a different way of paginating results.

bwaidelich commented on Neos.Neos/Classes/Controller/Service/AssetProxiesController.php:
```diff
@@ -0,0 +1,167 @@
+                    $assetProxyQueryResult->next();
+                    $addedResults++;
+                }
+            } catch (\Exception $exception) {
```
catching the base exception is quite dangerous. Is that really required here?

robertlemke replied:
We could log the exception, but there is no meaningful way to display the error to the user at this point.

robertlemke commented:
Thanks for your review, @bwaidelich. In practice, I don't share your concerns. I tested this code with 3 asset sources enabled, one of them containing ten thousands of assets, and did not experience speed problems or killed user experiences so far.

I agree that it would be nice to have type ahead options configurable for asset sources which may have a problem with this way of searching, but I'd like to see that problem happening first.

bwaidelich commented:
> I [...] did not experience speed problems or killed user experiences so far

But @daniellienert did. In the meantime there are quite some asset source implementations out there some of which are not as performant and/or have rate limits applied

bwaidelich commented:
I'll check if I can find an easy way around later today and will comment my findings here!

robertlemke commented:
Regarding speed: please consider my comment regarding pagination. And also, that the asset sources API is not officially released yet. We may come up with breaking changes in the API which improves performance, or implementors need to change their asset source in order to provide better performance.

bwaidelich commented:
So I played around with this a little more and IMO it is a great improvement already and could go in.
I do have a couple of minor issues that we could address (maybe in a follow up):

### Configurable asset sources

The asset-proxies endpoint already supports filtering by `assetSourceIdentifier`. I didn't figure out where this is used currently but IMO it would be nice if that would accept an array and we could go for something like
```yaml
editorOptions:
  allowedAssetSources: ['neos', 'unsplash']
```

### Global limit for the suggestions

Currently the limit is enforced **per asset source**. If there are multiple configured, this can lead to a very long list of suggestions that "break" the UI (see below). We could consider applying the limit to the complete list instead, or maybe more sensible even: spread the limit over all asset sources evenly so that the total result never exceeds the limit.

### Make the asset source visible

Currently the list of suggestions is not only (possibly) long but also there is no way to tell which asset comes from where:

![image](https://user-images.githubusercontent.com/307571/46742800-0da09580-cca8-11e8-953a-386e650e5d58.png)

I did a mini experiment (just a [2-liner](https://gist.github.com/bwaidelich/9279334ba869e6e6617e9374d44f198b/revisions#diff-168726dbe96b3ce427e7fedce31bb0bc)) to add the asset sources as groups.
This needs some further thoughts, but it's already a bit more "tidy" IMO:

![image](https://user-images.githubusercontent.com/307571/46743068-928baf00-cca8-11e8-93b9-acb734be47a6.png)

kdambekalns commented:
Thanks for trying that out, @bwaidelich. I think those could go into separate PRs, as you suggested. Would you open issues for them, to keep track?

But… about the grouping by asset source, if that's a two-liner now, maybe that could be done right away. What further thoughts would that need? Immediately I see only a need for assigning a "proper name" for a source…

bwaidelich commented:
> What further thoughts would that need?

The endpoint currently only returns the id and IIRC there is no notion of an "asset source _name_" yet. Also the grouping should only happen if there are more than one (enabled) source

kdambekalns commented:
> The endpoint currently only returns the id

Well, but you solved that in the two-liner, no?

And yes, the _name_ is obvious.

robertlemke commented:
Allright, thanks for your reviews, so far!

I'm also all in for further improvements of the UI, for example displaying additional metadata for a selected (but not applied) asset, so you can distinguish assets like "flownative-brand-2018-logo-cropped-tight-whitespace-800x200px.png" and "flownative-brand-2018-logo-cropped-tight-whitespace-200x70px.png".

Two questions:
- should we merge this now and provide further improvements in follow-up pull requests?
- which branch should this be merged into?

For the latter: This is basically a new feature, so it should just go into the next release. However, you could also say that it's fixing the Asset Editor, so it should go into earlier versions as well. The only real reason for me to provide this for 4.0 (because we introduced asset sources there) and 4.1 would be that the Neos UI package otherwise would need a feature flag, because it needs to be compatible with all 4.x versions.

bwaidelich commented:
In my opinion the two questions interact a little:
I think it's totally fine to adjust this in subsequent PRs but if we want to have this in released branches (which I definitely do) we should maybe at least add that "assetSources" backend option so that the 3rd party lookup can be "disabled" if needed. The endpoint already supports this (partly: it only allows one `$assetSourceIdentifier`  and this should support arrays IMO) so it's "only" a matter of forwarding that option in the component request.

@daniellienert what is your opinion?

robertlemke commented:
> we should maybe at least add that "assetSources" backend option so that the 3rd party lookups can be "disabled" if needed.

Okay, I'll implement this now. If that's working fine, we should merge this though and add additional features into follow-ups.

robertlemke commented:
I implemented the following new features:

- the result limit for the Asset Proxy Service is now specified through the _limit_ parameter and counts as a total result limit across all asset sources which deliver results
- results are grouped by asset source and the limit is applied evenly across all sources
- for the user interface the asset source label is used instead of the identifier
- exceptions of an asset source are not caught anymore, but instead the Asset Proxy Service relies on the query result's _valid()_ method to return false if there are any problems

<img width="318" alt="image" src="https://user-images.githubusercontent.com/95582/47486198-15337300-d840-11e8-84f4-484335dd7118.png">

bwaidelich commented:
@robertlemke you are the man! testing asap

bwaidelich commented on Neos.Media/Classes/Domain/Service/AssetSourceService.php:
```diff
@@ -91,7 +91,7 @@
      * @return AssetSourceInterface[]
      */
-    public function getAssetsSources(): array
+    public function getAssetSources(): array
```
oops, good catch

bwaidelich commented on Neos.Neos/Classes/Controller/Service/AssetProxiesController.php:
```diff
@@ -0,0 +1,173 @@
+                    $assetProxiesByAssetSource[$assetProxy->getAssetSource()->getIdentifier()][] = $assetProxy;
+                    $totalAddedResults ++;
+                }
+                if ($totalAddedResults === $limit) {
```
nitpick: that change can be moved inside the if condition above as that's the only part that alters the counter
Or even safe a nesting level by doing
```php
foreach ($assetProxies as $assetProxy) {
    if (!$assetProxy instanceof AssetProxyInterface) {
        continue;
    }
    $assetProxiesByAssetSource[$assetProxy->getAssetSource()->getIdentifier()][] = $assetProxy;
    if (++$totalAddedResults === $limit) {
        break 2;
    }
}
```
but I know that you dislike `continue` :)

robertlemke commented:
> Also, what about the assetSources backendOption mentioned above (#2200 (comment))?

I'm not completely sure of that option yet. What I can imagine is introducing _editor_ options which allow to disable certain asset sources for an editor. But I would not like an option you can set for the individual asset source, if that is what you meant. Let's skip this feature for now.

> ++$totalAddedResults === $limit
Actually I don't like that kind of code … ;-)

bwaidelich commented:
FYI: When searching for a term that does not yield results I currently get a 500 from the endpoint (triggered by the `dl/assetsource-mediawiki` client). But it seems like this is due to a bug in the implementation (see https://github.com/daniellienert/assetsource-mediawiki/issues/4)

bwaidelich commented:
@robertlemke I think my comment above (https://github.com/neos/neos-development-collection/pull/2200#issuecomment-432997896) got swept under the carpet :)
